File: app/src/main/java/org/blitzortung/android/app/view/AlertView.java
package org.blitzortung.android.app.view;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Paint.Align;
import android.graphics.Paint.Style;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RectF;
import android.location.Location;
import android.util.AttributeSet;
import android.view.View;

import com.annimon.stream.function.Consumer;

import org.blitzortung.android.alert.AlertParameters;
import org.blitzortung.android.alert.data.AlertSector;
import org.blitzortung.android.alert.data.AlertSectorRange;
import org.blitzortung.android.alert.data.AlertStatus;
import org.blitzortung.android.alert.event.AlertEvent;
import org.blitzortung.android.alert.event.AlertResultEvent;
import org.blitzortung.android.app.R;
import org.blitzortung.android.app.helper.ViewHelper;
import org.blitzortung.android.location.LocationEvent;
import org.blitzortung.android.map.overlay.color.ColorHandler;

import java.util.List;

public class AlertView extends View {

    private static final int TEXT_MINIMUM_SIZE = 300;
    private static final int DEFAULT_FONT_SIZE = 20;
    private static final PorterDuffXfermode XFERMODE_CLEAR = new PorterDuffXfermode(PorterDuff.Mode.CLEAR);
    private static final PorterDuffXfermode XFERMODE_SRC = new PorterDuffXfermode(PorterDuff.Mode.SRC);
    private final RectF arcArea = new RectF();
    private final Paint background = new Paint();
    private final Paint sectorPaint = new Paint();
    private final Paint lines = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint textStyle = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint warnText = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint transfer = new Paint();
    private final String[] alarmNotAvailableTextLines;
    private ColorHandler colorHandler;
    private int intervalDuration;
    private Bitmap temporaryBitmap;
    private Canvas temporaryCanvas;
    private AlertStatus alertStatus;
    private Consumer<AlertEvent> alertEventConsumer = new Consumer<AlertEvent>() {
        @Override
        public void accept(AlertEvent event) {
            if (event instanceof AlertResultEvent) {
                AlertResultEvent alertResultEvent = (AlertResultEvent) event;

                alertStatus = alertResultEvent.getAlertStatus();
            } else {
                alertStatus = null;
            }
            invalidate();
        }
    };
    private Consumer<LocationEvent> locationEventConsumer = locationEvent -> {
        final Location location = locationEvent.getLocation();
        final int visibility = location != null ? View.VISIBLE : View.INVISIBLE;
        setVisibility(visibility);
        invalidate();
    };

    @SuppressWarnings("unused")
    public AlertView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    @SuppressWarnings("unused")
    public AlertView(Context context) {
        this(context, null, 0);
    }

    public AlertView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        alarmNotAvailableTextLines = context.getString(R.string.alarms_not_available).split("\n");

        lines.setColor(0xff404040);
        lines.setStyle(Style.STROKE);

        textStyle.setColor(0xff404040);
        textStyle.setTextSize(ViewHelper.pxFromSp(this, 10));

        background.setColor(0xffb0b0b0);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int parentWidth = MeasureSpec.getSize(widthMeasureSpec);
        int parentHeight = MeasureSpec.getSize(heightMeasureSpec);

        int size = Math.min(parentWidth, parentHeight);

        super.onMeasure(MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY), MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY));
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        int size = Math.max(getWidth(), getHeight());
        int pad = 4;

        float center = size / 2.0f;
        float radius = center - pad;

        prepareTemporaryBitmap(size);

        if (alertStatus != null && intervalDuration != 0) {
            AlertParameters alertParameters = alertStatus.getAlertParameters();
            final float[] rangeSteps = alertParameters.getRangeSteps();
            final int rangeStepCount = rangeSteps.length;
            final float radiusIncrement = radius / rangeStepCount;
            final float sectorWidth = 360 / alertParameters.getSectorLabels().length;

            lines.setColor(colorHandler.getLineColor());
            lines.setStrokeWidth(size / 150);

            textStyle.setTextAlign(Align.CENTER);
            textStyle.setColor(colorHandler.getTextColor());

            long actualTime = System.currentTimeMillis();

            for (AlertSector alertSector : alertStatus.getSectors()) {

                float startAngle = alertSector.getMinimumSectorBearing() + 90 + 180;

                final List<AlertSectorRange> ranges = alertSector.getRanges();
                for (int rangeIndex = ranges.size() - 1; rangeIndex >= 0; rangeIndex--) {
                    AlertSectorRange alertSectorRange = ranges.get(rangeIndex);

                    float sectorRadius = (rangeIndex + 1) * radiusIncrement;
                    float leftTop = center - sectorRadius;
                    float bottomRight = center + sectorRadius;

                    boolean drawColor = alertSectorRange.getStrikeCount() > 0;
                    if (drawColor) {
                        final int color = colorHandler.getColor(actualTime, alertSectorRange.getLatestStrikeTimestamp(), intervalDuration);
                        sectorPaint.setColor(color);
                    }
                    arcArea.set(leftTop, leftTop, bottomRight, bottomRight);
                    temporaryCanvas.drawArc(arcArea, startAngle, sectorWidth, true, drawColor ? sectorPaint : background);
                }
            }

            for (AlertSector alertSector : alertStatus.getSectors()) {
                double bearing = alertSector.getMinimumSectorBearing();
                temporaryCanvas.drawLine(center, center, center + (float) (radius * Math.sin(bearing / 180.0f * Math.PI)), center
                        + (float) (radius * -Math.cos(bearing / 180.0f * Math.PI)), lines);

                if (size > TEXT_MINIMUM_SIZE) {
                    drawSectorLabel(center, radiusIncrement, alertSector, bearing + sectorWidth / 2.0);
                }
            }

            textStyle.setTextAlign(Align.RIGHT);
            float textHeight = textStyle.getFontMetrics(null);
            for (int radiusIndex = 0; radiusIndex < rangeStepCount; radiusIndex++) {
                float leftTop = center - (radiusIndex + 1) * radiusIncrement;
                float bottomRight = center + (radiusIndex + 1) * radiusIncrement;
                arcArea.set(leftTop, leftTop, bottomRight, bottomRight);
                temporaryCanvas.drawArc(arcArea, 0, 360, false, lines);

                if (size > TEXT_MINIMUM_SIZE) {
                    String text = String.format("%.0f", rangeSteps[radiusIndex]);
                    temporaryCanvas.drawText(text, center + (radiusIndex + 0.85f) * radiusIncrement, center
                            + textHeight / 3f, textStyle);
                    if (radiusIndex == rangeStepCount - 1) {
                        temporaryCanvas.drawText(alertParameters.getMeasurementSystem().getUnitName(), center + (radiusIndex + 0.85f) * radiusIncrement, center
                                + textHeight * 1.33f, textStyle);
                    }
                }
            }

        } else if (size > TEXT_MINIMUM_SIZE) {

            warnText.setColor(0xffa00000);
            warnText.setTextAlign(Align.CENTER);
            warnText.setTextSize(DEFAULT_FONT_SIZE);

            //Find the smallest possible scale
            float scale = 1;
            for (String alarmNotAvailableTextLine : alarmNotAvailableTextLines) {
                //Use margin of 20
                //Because text is centered, margin left and right will be 10
                scale = Math.min(scale, getWidth() - 20) / warnText.measureText(alarmNotAvailableTextLine);
            }

            //Now scale the text so we can use the whole width of the canvas
            warnText.setTextSize(scale * DEFAULT_FONT_SIZE);

            for (int line = 0; line < alarmNotAvailableTextLines.length; line++) {
                temporaryCanvas.drawText(alarmNotAvailableTextLines[line], center, center + (line - 1) * warnText.getFontMetrics(null), warnText);
            }
        }
        canvas.drawBitmap(temporaryBitmap, 0, 0, transfer);
    }

    private void drawSectorLabel(float center, float radiusIncrement, AlertSector sector, double bearing) {
        if (bearing != 90.0) {
            final String text = sector.getLabel();
            float textRadius = (sector.getRanges().size() - 0.5f) * radiusIncrement;
            temporaryCanvas.drawText(text, center
                    + (float) (textRadius * Math.sin(bearing / 180.0 * Math.PI)), center
                    + (float) (textRadius * -Math.cos(bearing / 180.0 * Math.PI)) + textStyle.getFontMetrics(null) / 3f, textStyle);
        }
    }

    private void prepareTemporaryBitmap(int size) {
        if (temporaryBitmap == null) {
            temporaryBitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);
            temporaryCanvas = new Canvas(temporaryBitmap);
        }
        background.setColor(colorHandler.getBackgroundColor());
        background.setXfermode(XFERMODE_CLEAR);
        temporaryCanvas.drawPaint(background);

        background.setXfermode(XFERMODE_SRC);
    }

    public Consumer<AlertEvent> getAlertEventConsumer() {
        return alertEventConsumer;
    }

    public Consumer<LocationEvent> getLocationEventConsumer() {
        return locationEventConsumer;
    }

    public void setColorHandler(ColorHandler colorHandler, int intervalDuration) {
        this.colorHandler = colorHandler;
        this.intervalDuration = intervalDuration;
    }

    @Override
    public void setBackgroundColor(int backgroundColor) {
        background.setColor(backgroundColor);
    }

    public void setAlpha(int alpha) {
        transfer.setAlpha(alpha);
    }

}

File: app/src/main/java/org/blitzortung/android/app/view/AlertView.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.app.view

import android.content.Context
import android.graphics.*
import android.graphics.Paint.Align
import android.graphics.Paint.Style
import android.location.Location
import android.preference.PreferenceManager
import android.util.AttributeSet
import android.util.Log
import org.blitzortung.android.alert.AlertResult
import org.blitzortung.android.alert.data.AlertSector
import org.blitzortung.android.alert.event.AlertEvent
import org.blitzortung.android.alert.event.AlertResultEvent
import org.blitzortung.android.alert.handler.AlertHandler
import org.blitzortung.android.app.Main
import org.blitzortung.android.app.R
import org.blitzortung.android.data.MainDataHandler
import org.blitzortung.android.dialogs.AlertDialog
import org.blitzortung.android.dialogs.AlertDialogColorHandler
import org.blitzortung.android.location.LocationEvent
import org.blitzortung.android.map.overlay.color.ColorHandler
import org.blitzortung.android.util.TabletAwareView
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

class AlertView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyle: Int = 0
) : TabletAwareView(context, attrs, defStyle) {
    private val arcArea = RectF()
    private val background = Paint()
    private val sectorPaint = Paint()
    private val lines = Paint(Paint.ANTI_ALIAS_FLAG)
    private val textStyle = Paint(Paint.ANTI_ALIAS_FLAG)
    private val warnText = Paint(Paint.ANTI_ALIAS_FLAG)
    private val transfer = Paint()
    private val alarmNotAvailableTextLines: Array<String> = context.getString(R.string.alarms_not_available)
            .split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    private lateinit var colorHandler: ColorHandler
    private var intervalDuration: Int = 0
    private var temporaryBitmap: Bitmap? = null
    private var temporaryCanvas: Canvas? = null
    private var alertResult: AlertResult? = null
    private var location: Location? = null
    private var enableDescriptionText = false

    val alertEventConsumer: (AlertEvent?) -> Unit = { event ->
        Log.v(Main.LOG_TAG, "AlertView alertEventConsumer received $event")
        alertResult = if (event is AlertResultEvent) {
            event.alertResult
        } else {
            null
        }
        invalidate()
    }

    val locationEventConsumer: (LocationEvent) -> Unit = { locationEvent ->
        Log.v(Main.LOG_TAG, "AlertView received location ${locationEvent.location}")
        location = locationEvent.location
        val visibility = if (location != null) VISIBLE else INVISIBLE
        setVisibility(visibility)
        invalidate()
    }

    init {
        with(lines) {
            color = 0xff404040.toInt()
            style = Style.STROKE
        }

        with(textStyle) {
            color = 0xff404040.toInt()
            textSize = 0.8f * this@AlertView.textSize * textSizeFactor(context)
        }

        background.color = 0xffb0b0b0.toInt()
    }

    fun enableLongClickListener(dataHandler: MainDataHandler, alertHandler: AlertHandler) {

        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

        setOnLongClickListener {
            AlertDialog(context, AlertDialogColorHandler(sharedPreferences), dataHandler, alertHandler)
                    .show()

            true
        }
    }

    fun enableDescriptionText() {
        enableDescriptionText = true
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val getSize = fun(spec: Int) = MeasureSpec.getSize(spec)

        val parentWidth = getSize(widthMeasureSpec) * sizeFactor
        val parentHeight = getSize(heightMeasureSpec) * sizeFactor

        val size = min(parentWidth.toInt(), parentHeight.toInt())

        super.onMeasure(MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY), MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY))
    }

    override fun onDraw(canvas: Canvas) {
        val size = max(width, height)
        val pad = 4

        val center = size / 2.0f
        val radius = center - pad

        prepareTemporaryBitmap(size)

        val alertResult = alertResult
        val temporaryCanvas = temporaryCanvas
        val temporaryBitmap = temporaryBitmap
        if (temporaryBitmap != null && temporaryCanvas != null) {
            if (alertResult != null && intervalDuration != 0) {
                val alertParameters = alertResult.parameters
                val rangeSteps = alertParameters.rangeSteps
                val rangeStepCount = rangeSteps.size
                val radiusIncrement = radius / rangeStepCount
                val sectorWidth = (360 / alertParameters.sectorLabels.size).toFloat()

                with(lines) {
                    color = colorHandler.lineColor
                    strokeWidth = (size / 150).toFloat()
                }

                with(textStyle) {
                    textAlign = Align.CENTER
                    color = colorHandler.textColor
                }

                val actualTime = System.currentTimeMillis()

                for (alertSector in alertResult.sectors) {

                    val startAngle = alertSector.minimumSectorBearing + 90f + 180f

                    val ranges = alertSector.ranges
                    for (rangeIndex in ranges.indices.reversed()) {
                        val alertSectorRange = ranges[rangeIndex]

                        val sectorRadius = (rangeIndex + 1) * radiusIncrement
                        val leftTop = center - sectorRadius
                        val bottomRight = center + sectorRadius

                        val drawColor = alertSectorRange.strikeCount > 0
                        if (drawColor) {
                            val color = colorHandler.getColor(actualTime, alertSectorRange.latestStrikeTimestamp, intervalDuration)
                            sectorPaint.color = color
                        }
                        arcArea.set(leftTop, leftTop, bottomRight, bottomRight)
                        temporaryCanvas.drawArc(arcArea, startAngle, sectorWidth, true, if (drawColor) sectorPaint else background)
                    }
                }

                for (alertSector in alertResult.sectors) {
                    val bearing = alertSector.minimumSectorBearing.toDouble()
                    temporaryCanvas.drawLine(center, center, center + (radius * sin(bearing / 180.0f * Math.PI)).toFloat(), center + (radius * -cos(bearing / 180.0f * Math.PI)).toFloat(), lines)

                    if (enableDescriptionText && size > TEXT_MINIMUM_SIZE) {
                        drawSectorLabel(center, radiusIncrement, alertSector, bearing + sectorWidth / 2.0)
                    }
                }

                textStyle.textAlign = Align.RIGHT
                val textHeight = textStyle.getFontMetrics(null)
                for (radiusIndex in 0 until rangeStepCount) {
                    val leftTop = center - (radiusIndex + 1) * radiusIncrement
                    val bottomRight = center + (radiusIndex + 1) * radiusIncrement
                    arcArea.set(leftTop, leftTop, bottomRight, bottomRight)
                    temporaryCanvas.drawArc(arcArea, 0f, 360f, false, lines)

                    if (enableDescriptionText && size > TEXT_MINIMUM_SIZE) {
                        val text = "%.0f".format(rangeSteps[radiusIndex])
                        temporaryCanvas.drawText(text, center + (radiusIndex + 0.85f) * radiusIncrement, center + textHeight / 3f, textStyle)
                        if (radiusIndex == rangeStepCount - 1) {
                            val distanceUnit = resources.getString(alertParameters.measurementSystem.unitNameString)
                            temporaryCanvas.drawText(distanceUnit, center + (radiusIndex + 0.85f) * radiusIncrement, center + textHeight * 1.33f, textStyle)
                        }
                    }
                }

            } else {
                if (enableDescriptionText && size > TEXT_MINIMUM_SIZE) {
                    drawAlertOrLocationMissingMessage(center, temporaryCanvas)
                } else {
                    if (location != null) {
                        drawOwnLocationSymbol(center, radius, size, temporaryCanvas)
                    }
                }
            }
            canvas.drawBitmap(temporaryBitmap, 0f, 0f, transfer)
        }
    }

    private fun drawAlertOrLocationMissingMessage(center: Float, canvas: Canvas) {
        with(warnText) {
            color = resources.getColor(R.color.text_warning)
            textAlign = Align.CENTER
            textSize = DEFAULT_FONT_SIZE.toFloat()

            val maxWidth = alarmNotAvailableTextLines.map { warnText.measureText(it) }.max()
                    ?: width.toFloat() - 20
            val scale = (width - 20).toFloat() / maxWidth

            //Now scale the text so we can use the whole width of the canvas
            textSize = scale * DEFAULT_FONT_SIZE
        }


        for (line in alarmNotAvailableTextLines.indices) {
            canvas.drawText(alarmNotAvailableTextLines[line], center, center + (line - 1) * warnText.getFontMetrics(null), warnText)
        }
    }

    private fun drawOwnLocationSymbol(center: Float, radius: Float, size: Int, temporaryCanvas: Canvas) {
        with(lines) {
            color = colorHandler.lineColor
            strokeWidth = (size / 80).toFloat()
        }

        val largeRadius = radius * 0.8f
        val leftTop = center - largeRadius
        val bottomRight = center + largeRadius
        arcArea.set(leftTop, leftTop, bottomRight, bottomRight)
        temporaryCanvas.drawArc(arcArea, 0f, 360f, false, lines)

        val smallRadius = radius * 0.6f
        temporaryCanvas.drawLine(center - smallRadius, center, center + smallRadius, center, lines)
        temporaryCanvas.drawLine(center, center - smallRadius, center, center + smallRadius, lines)
    }

    private fun drawSectorLabel(center: Float, radiusIncrement: Float, sector: AlertSector, bearing: Double) {
        if (bearing != 90.0) {
            val text = sector.label
            val textRadius = (sector.ranges.size - 0.5f) * radiusIncrement
            temporaryCanvas!!.drawText(text, center + (textRadius * sin(bearing / 180.0 * Math.PI)).toFloat(), center + (textRadius * -cos(bearing / 180.0 * Math.PI)).toFloat() + textStyle.getFontMetrics(null) / 3f, textStyle)
        }
    }

    private fun prepareTemporaryBitmap(size: Int) {
        if (temporaryBitmap == null) {
            val temporaryBitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
            this.temporaryBitmap = temporaryBitmap
            temporaryCanvas = Canvas(temporaryBitmap)
        }
        background.color = colorHandler.backgroundColor
        background.xfermode = XFERMODE_CLEAR
        temporaryCanvas!!.drawPaint(background)

        background.xfermode = XFERMODE_SRC
    }

    fun setColorHandler(colorHandler: ColorHandler, intervalDuration: Int) {
        this.colorHandler = colorHandler
        this.intervalDuration = intervalDuration
    }

    override fun setBackgroundColor(backgroundColor: Int) {
        background.color = backgroundColor
    }

    fun setAlpha(alpha: Int) {
        transfer.alpha = alpha
    }

    companion object {
        private const val TEXT_MINIMUM_SIZE = 300
        private const val DEFAULT_FONT_SIZE = 20
        private val XFERMODE_CLEAR = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
        private val XFERMODE_SRC = PorterDuffXfermode(PorterDuff.Mode.SRC)
    }

}
